packages/trpc/server/template-router/router.ts

```diff
@@ -0,0 +1,116 @@
+import { TRPCError } from '@trpc/server';
+
+import { createDocumentFromTemplate } from '@documenso/lib/server-only/template/create-document-from-template';
+import { createTemplate } from '@documenso/lib/server-only/template/create-template';
+import { createTemplateFromDocument } from '@documenso/lib/server-only/template/create-template-from-document';
+import { deleteTemplate } from '@documenso/lib/server-only/template/delete-template';
+import { duplicateTemplate } from '@documenso/lib/server-only/template/duplicate-template';
+
+import { authenticatedProcedure, router } from '../trpc';
+import {
+  ZCreateDocumentFromTemplateMutationSchema,
+  ZCreateTemplateFromDocumentMutationSchema,
+  ZCreateTemplateMutationSchema,
+  ZDeleteTemplateMutationSchema,
+  ZDuplicateTemplateMutationSchema,
+} from './schema';
+
+export const templateRouter = router({
+  createTemplate: authenticatedProcedure
+    .input(ZCreateTemplateMutationSchema)
+    .mutation(async ({ input, ctx }) => {
+      try {
+        const { title, templateDocumentDataId } = input;
+
+        return await createTemplate({
+          title,
+          userId: ctx.user.id,
+          templateDocumentDataId,
+        });
+      } catch (err) {
+        console.error(err);
+
+        throw new TRPCError({
+          code: 'BAD_REQUEST',
+          message: 'We were unable to create this template. Please try again later.',
+        });
+      }
+    }),
+
+  createDocumentFromTemplate: authenticatedProcedure
+    .input(ZCreateDocumentFromTemplateMutationSchema)
+    .mutation(async ({ input, ctx }) => {
+      try {
+        const { templateId } = input;
+
+        return await createDocumentFromTemplate({
+          templateId,
+          userId: ctx.user.id,
+        });
+      } catch (err) {
+        throw new TRPCError({
+          code: 'BAD_REQUEST',
+          message: 'We were unable to create this document. Please try again later.',
+        });
+      }
+    }),
+
+  createTemplateFromDocument: authenticatedProcedure
+    .input(ZCreateTemplateFromDocumentMutationSchema)
+    .mutation(async ({ input, ctx }) => {
+      try {
+        const { documentId } = input;
+
+        return await createTemplateFromDocument({
+          documentId,
+          userId: ctx.user.id,
+        });
+      } catch (err) {
+        console.error(err);
+
+        throw new TRPCError({
+          code: 'BAD_REQUEST',
+          message: 'We were unable to create this template. Please try again later.',
+        });
+      }
+    }),
+
+  duplicateTemplate: authenticatedProcedure
```

Comment:
What would be the context for wanting to duplicate a template? It's a nice thing to have implemented but do we think it's required? 🤔

Reply:
To create variations of a base template without starting from scratch each time was what I had in mind then

Reply:
Yeah feels relevant, can we update the duplicate to behave a bit like the one on the documents page with the dialog? It took me by surprise when it just went and did it without confirmation 😄